engine/consensus/matching/core.go

```diff
@@ -562,222 +559,244 @@ func (c *Core) CheckSealing() error {
 // collected enough approvals on a per-chunk basis, as defined by the matchChunk
 // function. It also filters out results that have an incorrect sub-graph.
 // It specifically returns the information for the next unsealed results which will
 // be useful for debugging the potential sealing halt issue
-func (c *Core) sealableResults() ([]*flow.IncorporatedResult, nextUnsealedResults, error) {
-	var results []*flow.IncorporatedResult
+func (c *Core) sealableResults() ([]*flow.IncorporatedResult, *sealingtracker.SealingTracker, error) {
+	//tracking functionality to collection information about the sealing status.
+	sealingTracker := sealingtracker.NewSealingTracker(c.state)
 
 	lastFinalized, err := c.state.Final().Head()
 	if err != nil {
 		return nil, nil, fmt.Errorf("failed to get last finalized block: %w", err)
 	}
 
-	nextUnsealed, nextUnsealedIsFinalized, err := nextUnsealedID(c.state)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get the next unsealed block id: %w", err)
-	}
-
-	nextUnsealeds := make([]*nextUnsealedResult, 0)
-	// go through the results mempool and check which ones we have collected
-	// enough approvals for
+	// go through the results mempool and check which ones we can construct a candidate seal for
+	var results []*flow.IncorporatedResult
 	for _, incorporatedResult := range c.incorporatedResults.All() {
-		// not finding the block header for an incorporated result is a fatal
-		// implementation bug, as we only add results to the IncorporatedResults
-		// mempool, where _both_ the block that incorporates the result as well
-		// as the block the result pertains to are known
-		block, err := c.headersDB.ByBlockID(incorporatedResult.Result.BlockID)
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not retrieve block: %w", err)
-		}
-
-		// At this point we can be sure that all needed checks on validity of ER
-		// were executed prior to this point, since we perform validation of every ER
-		// before adding it into mempool. Hence, the mempool can contain only valid entries.
-
-		// the chunk assigment is based on the first block in its fork which
-		// contains a receipt that commits to this result.
-		assignment, err := c.assigner.Assign(incorporatedResult.Result, incorporatedResult.IncorporatedBlockID)
+		// Can we seal following the happy-path protocol?
+		sealingStatus, err := c.hasEnoughApprovals(incorporatedResult, sealingTracker)
 		if state.IsNoValidChildBlockError(err) {
 			continue
 		}
 		if err != nil {
-			// at this point, we know the block and a valid child block exists. Not being able to compute
-			// the assignment constitutes a fatal implementation bug:
-			return nil, nil, fmt.Errorf("could not determine chunk assignment: %w", err)
+			return nil, nil, fmt.Errorf("internal error matching chunk approvals to incorporated result: %w", err)
 		}
+		happySealable := sealingStatus.SufficientApprovalsForSealing
```

Comment:
"happy sealable" or "happy path sealable" are confusing. Maybe just call it `sealableWithEnoughApprovals`.